File: backend/internal/db/postgres/training_plan.go
package postgres

import (
	"context"
	"fmt"

	"github.com/anish-chanda/cadent/backend/internal/models"
	"github.com/jackc/pgx/v5"
)

func (s *PostgresDB) GetTrainingPlans(ctx context.Context, searchQuery string, sport string) ([]models.TrainingPlan, error) {
	query := `
		SELECT id, created_by_user_id, title, description, primary_sport, difficulty,
			   duration_weeks, recommended_workouts_per_week, is_system, created_at, updated_at
		FROM training_plans
		WHERE 1=1
	`
	args := []interface{}{}
	argIdx := 1

	if searchQuery != "" {
		query += fmt.Sprintf(" AND (title ILIKE $%d OR description ILIKE $%d)", argIdx, argIdx)
		args = append(args, "%"+searchQuery+"%")
		argIdx++
	}
	if sport != "" {
		query += fmt.Sprintf(" AND primary_sport = $%d", argIdx)
		args = append(args, sport)
		argIdx++
	}

	query += " ORDER BY is_system DESC, created_at DESC"

	rows, err := s.pool.Query(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("failed to get training plans: %w", err)
	}
	defer rows.Close()

	var plans []models.TrainingPlan
	for rows.Next() {
		var p models.TrainingPlan
		if err := rows.Scan(
			&p.ID, &p.CreatedByUserID, &p.Title, &p.Description, &p.PrimarySport, &p.Difficulty,
			&p.DurationWeeks, &p.RecommendedWorkoutsPerWeek, &p.IsSystem, &p.CreatedAt, &p.UpdatedAt,
		); err != nil {
			return nil, fmt.Errorf("failed to scan training plan: %w", err)
		}
		plans = append(plans, p)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iteration error: %w", err)
	}

	if plans == nil {
		plans = []models.TrainingPlan{}
	}

	return plans, nil
}

func (s *PostgresDB) GetTrainingPlanByID(ctx context.Context, planID string) (*models.TrainingPlan, error) {
	query := `
		SELECT id, created_by_user_id, title, description, primary_sport, difficulty,
			   duration_weeks, recommended_workouts_per_week, is_system, created_at, updated_at
		FROM training_plans
		WHERE id = $1
	`

	var plan models.TrainingPlan
	err := s.pool.QueryRow(ctx, query, planID).Scan(
		&plan.ID, &plan.CreatedByUserID, &plan.Title, &plan.Description, &plan.PrimarySport, &plan.Difficulty,
		&plan.DurationWeeks, &plan.RecommendedWorkoutsPerWeek, &plan.IsSystem, &plan.CreatedAt, &plan.UpdatedAt,
	)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("failed to get training plan by id: %w", err)
	}

	return &plan, nil
}

func (s *PostgresDB) GetTrainingPlanWorkouts(ctx context.Context, planID string) ([]models.TrainingPlanWorkout, error) {
	query := `
		SELECT id, training_plan_id, sequence_index, template_day_offset, type, title, description,
			   planned_distance_m, planned_duration_s, planned_elevation_gain_m, target_avg_speed_mps,
			   target_power_watt, created_at, updated_at
		FROM training_plan_workouts
		WHERE training_plan_id = $1
		ORDER BY sequence_index ASC
	`

	rows, err := s.pool.Query(ctx, query, planID)
	if err != nil {
		return nil, fmt.Errorf("failed to query workouts: %w", err)
	}
	defer rows.Close()

	var workouts []models.TrainingPlanWorkout
	for rows.Next() {
		var w models.TrainingPlanWorkout
		if err := rows.Scan(
			&w.ID, &w.TrainingPlanID, &w.SequenceIndex, &w.TemplateDayOffset, &w.Type, &w.Title,
			&w.Description, &w.PlannedDistanceM, &w.PlannedDurationS, &w.PlannedElevationGainM,
			&w.TargetAvgSpeedMps, &w.TargetPowerWatt, &w.CreatedAt, &w.UpdatedAt,
		); err != nil {
			return nil, fmt.Errorf("failed to scan workout: %w", err)
		}
		workouts = append(workouts, w)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iteration error: %w", err)
	}

	if workouts == nil {
		workouts = []models.TrainingPlanWorkout{}
	}

	return workouts, nil
}

func (s *PostgresDB) CreateUserTrainingPlanWithPlannedActivities(ctx context.Context, userPlan *models.UserTrainingPlan, plannedActivities []models.PlannedActivity) error {
	beginner, ok := s.pool.(interface {
		Begin(context.Context) (pgx.Tx, error)
	})
	if !ok {
		return fmt.Errorf("database pool does not support transactions")
	}

	tx, err := beginner.Begin(ctx)
	if err != nil {
		return fmt.Errorf("failed to begin transaction: %w", err)
	}

	committed := false
	defer func() {
		if !committed {
			_ = tx.Rollback(ctx)
		}
	}()

	insertUserPlanQuery := `
		INSERT INTO user_training_plans (
			user_id, training_plan_id, title, description, start_date, selected_workouts_per_week
		)
		VALUES ($1, $2, $3, $4, $5, $6)
		RETURNING id, created_at, updated_at
	`

	err = tx.QueryRow(ctx, insertUserPlanQuery,
		userPlan.UserID,
		userPlan.TrainingPlanID,
		userPlan.Title,
		userPlan.Description,
		userPlan.StartDate,
		userPlan.SelectedWorkoutsPerWeek,
	).Scan(&userPlan.ID, &userPlan.CreatedAt, &userPlan.UpdatedAt)
	if err != nil {
		return fmt.Errorf("failed to create user training plan: %w", err)
	}

	if len(plannedActivities) == 0 {
		if err := tx.Commit(ctx); err != nil {
			return fmt.Errorf("failed to commit transaction: %w", err)
		}
		committed = true
		return nil
	}

	batch := &pgx.Batch{}
	insertPlannedActivityQuery := `
		INSERT INTO planned_activities (
			user_id, title, description, type, start_time,
			planned_distance_m, planned_duration_s, planned_elevation_gain_m,
			target_avg_speed_mps, target_power_watt,
			user_training_plan_id, plan_sequence_index
		)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
	`

	for i := range plannedActivities {
		activity := plannedActivities[i]
		if activity.PlanSequenceIndex == nil {
			return fmt.Errorf("planned activity at index %d missing plan sequence index", i)
		}

		batch.Queue(
			insertPlannedActivityQuery,
			activity.UserID,
			activity.Title,
			activity.Description,
			activity.Type,
			activity.StartTime,
			activity.PlannedDistanceM,
			activity.PlannedDurationS,
			activity.PlannedElevationGainM,
			activity.TargetAvgSpeedMps,
			activity.TargetPowerWatt,
			userPlan.ID,
			*activity.PlanSequenceIndex,
		)
	}

	batchResults := tx.SendBatch(ctx, batch)
	for i := 0; i < len(plannedActivities); i++ {
		if _, err := batchResults.Exec(); err != nil {
			_ = batchResults.Close()
			return fmt.Errorf("failed to batch insert planned activities: %w", err)
		}
	}
	if err := batchResults.Close(); err != nil {
		return fmt.Errorf("failed to finalize planned activity batch: %w", err)
	}

	if err := tx.Commit(ctx); err != nil {
		return fmt.Errorf("failed to commit transaction: %w", err)
	}
	committed = true

	return nil
}

File: app/lib/models/training_plan.dart
enum TrainingPlanActivityTypeFilter { all, running, roadBiking }

extension TrainingPlanActivityTypeFilterX on TrainingPlanActivityTypeFilter {
  String get apiValue {
    switch (this) {
      case TrainingPlanActivityTypeFilter.all:
        return 'all';
      case TrainingPlanActivityTypeFilter.running:
        return 'running';
      case TrainingPlanActivityTypeFilter.roadBiking:
        return 'road_biking';
    }
  }

  String get label {
    switch (this) {
      case TrainingPlanActivityTypeFilter.all:
        return 'All';
      case TrainingPlanActivityTypeFilter.running:
        return 'Running';
      case TrainingPlanActivityTypeFilter.roadBiking:
        return 'Road Biking';
    }
  }
}

class TrainingPlan {
  final String id;
  final String title;
  final String? description;
  final String? primaryActivityType;
  final String difficulty;
  final int durationWeeks;
  final int recommendedWorkoutsPerWeek;
  final bool isSystem;
  final DateTime createdAt;
  final DateTime updatedAt;

  const TrainingPlan({
    required this.id,
    required this.title,
    this.description,
    this.primaryActivityType,
    required this.difficulty,
    required this.durationWeeks,
    required this.recommendedWorkoutsPerWeek,
    required this.isSystem,
    required this.createdAt,
    required this.updatedAt,
  });

  factory TrainingPlan.fromJson(Map<String, dynamic> json) {
    return TrainingPlan(
      id: json['id'] as String,
      title: (json['title'] as String?)?.trim().isNotEmpty == true
          ? json['title'] as String
          : 'Untitled Plan',
      description: json['description'] as String?,
      primaryActivityType: json['primary_activity_type'] as String?,
      difficulty: (json['difficulty'] as String?) ?? 'beginner',
      durationWeeks: (json['duration_weeks'] as num?)?.toInt() ?? 0,
      recommendedWorkoutsPerWeek:
          (json['recommended_workouts_per_week'] as num?)?.toInt() ?? 1,
      isSystem: json['is_system'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class TrainingPlanWorkout {
  final String id;
  final String trainingPlanId;
  final int sequenceIndex;
  final int templateDayOffset;
  final String type;
  final String title;
  final String? description;
  final double? plannedDistanceM;
  final int? plannedDurationS;
  final double? plannedElevationGainM;
  final double? targetAvgSpeedMps;
  final int? targetPowerWatt;
  final DateTime createdAt;
  final DateTime updatedAt;

  const TrainingPlanWorkout({
    required this.id,
    required this.trainingPlanId,
    required this.sequenceIndex,
    required this.templateDayOffset,
    required this.type,
    required this.title,
    this.description,
    this.plannedDistanceM,
    this.plannedDurationS,
    this.plannedElevationGainM,
    this.targetAvgSpeedMps,
    this.targetPowerWatt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory TrainingPlanWorkout.fromJson(Map<String, dynamic> json) {
    return TrainingPlanWorkout(
      id: json['id'] as String,
      trainingPlanId: json['training_plan_id'] as String,
      sequenceIndex: (json['sequence_index'] as num?)?.toInt() ?? 0,
      templateDayOffset: (json['template_day_offset'] as num?)?.toInt() ?? 0,
      type: (json['type'] as String?) ?? 'running',
      title: (json['title'] as String?)?.trim().isNotEmpty == true
          ? json['title'] as String
          : 'Workout',
      description: json['description'] as String?,
      plannedDistanceM: (json['planned_distance_m'] as num?)?.toDouble(),
      plannedDurationS: (json['planned_duration_s'] as num?)?.toInt(),
      plannedElevationGainM: (json['planned_elevation_gain_m'] as num?)
          ?.toDouble(),
      targetAvgSpeedMps: (json['target_avg_speed_mps'] as num?)?.toDouble(),
      targetPowerWatt: (json['target_power_watt'] as num?)?.toInt(),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class ImportTrainingPlanRequest {
  final DateTime startDate;
  final int selectedWorkoutsPerWeek;
  final String title;
  final String? description;

  const ImportTrainingPlanRequest({
    required this.startDate,
    required this.selectedWorkoutsPerWeek,
    required this.title,
    this.description,
  });

  Map<String, dynamic> toJson() {
    return {
      'startDate': startDate.toUtc().toIso8601String(),
      'selectedWorkoutsPerWeek': selectedWorkoutsPerWeek,
      'title': title,
      'description': description,
    };
  }
}

class ImportTrainingPlanDryRunRequest {
  final DateTime startDate;
  final int selectedWorkoutsPerWeek;
  final String? title;
  final String? description;

  const ImportTrainingPlanDryRunRequest({
    required this.startDate,
    required this.selectedWorkoutsPerWeek,
    this.title,
    this.description,
  });

  Map<String, dynamic> toJson() {
    return {
      'startDate': startDate.toUtc().toIso8601String(),
      'selectedWorkoutsPerWeek': selectedWorkoutsPerWeek,
      if (title != null) 'title': title,
      if (description != null) 'description': description,
    };
  }
}

class ImportTrainingPlanResponse {
  final String userTrainingPlanId;
  final int plannedActivitiesCreated;

  const ImportTrainingPlanResponse({
    required this.userTrainingPlanId,
    required this.plannedActivitiesCreated,
  });

  factory ImportTrainingPlanResponse.fromJson(Map<String, dynamic> json) {
    return ImportTrainingPlanResponse(
      userTrainingPlanId: json['userTrainingPlanId'] as String,
      plannedActivitiesCreated:
          (json['plannedActivitiesCreated'] as num?)?.toInt() ?? 0,
    );
  }
}

class TrainingPlanPreviewActivity {
  final String id;
  final String title;
  final String type;
  final DateTime startTime;

  const TrainingPlanPreviewActivity({
    required this.id,
    required this.title,
    required this.type,
    required this.startTime,
  });

  factory TrainingPlanPreviewActivity.fromJson(Map<String, dynamic> json) {
    return TrainingPlanPreviewActivity(
      id: json['id'] as String,
      title: (json['title'] as String?) ?? 'Activity',
      type: (json['type'] as String?) ?? 'running',
      startTime: DateTime.parse(json['start_time'] as String),
    );
  }
}

class TrainingPlanPreviewPlannedActivity {
  final String id;
  final String title;
  final String type;
  final DateTime startTime;
  final bool isDryRun;
  final String? matchedActivityId;

  const TrainingPlanPreviewPlannedActivity({
    required this.id,
    required this.title,
    required this.type,
    required this.startTime,
    required this.isDryRun,
    this.matchedActivityId,
  });

  factory TrainingPlanPreviewPlannedActivity.fromJson(
    Map<String, dynamic> json,
  ) {
    return TrainingPlanPreviewPlannedActivity(
      id: json['id'] as String,
      title: (json['title'] as String?) ?? 'Planned Activity',
      type: (json['type'] as String?) ?? 'running',
      startTime: DateTime.parse(json['start_time'] as String),
      isDryRun: json['is_dry_run'] as bool? ?? false,
      matchedActivityId: json['matched_activity_id'] as String?,
    );
  }
}

class TrainingPlanPreviewResponse {
  final List<TrainingPlanPreviewActivity> activities;
  final List<TrainingPlanPreviewPlannedActivity> plannedActivities;

  const TrainingPlanPreviewResponse({
    required this.activities,
    required this.plannedActivities,
  });

  factory TrainingPlanPreviewResponse.fromJson(Map<String, dynamic> json) {
    final activitiesRaw = (json['activities'] as List<dynamic>? ?? []);
    final plannedActivitiesRaw =
        (json['planned_activities'] as List<dynamic>? ?? []);

    return TrainingPlanPreviewResponse(
      activities: activitiesRaw
          .whereType<Map<String, dynamic>>()
          .map(TrainingPlanPreviewActivity.fromJson)
          .toList(),
      plannedActivities: plannedActivitiesRaw
          .whereType<Map<String, dynamic>>()
          .map(TrainingPlanPreviewPlannedActivity.fromJson)
          .toList(),
    );
  }
}
